src/gateway/sharding/shard_manager.rs

```diff
@@ -142,16 +142,29 @@ impl ShardManager {
     ) -> Result<(), GatewayError> {
         self.initialize(shard_index, shard_init, shard_total);
         loop {
-            if let Ok(Some(msg)) =
-                timeout(self.wait_time_between_shard_start, self.manager_rx.next()).await
-            {
+            let batch = self.queue.pop_batch();
+            let msg = if batch.is_empty() {
+                // This function is the only code that can add new shards
+                // to start to the queue directly (enforced by `&mut`), so
+                // if the batch is empty, it will always be empty until a
+                // `ShardManagerMessage::Boot` is received here.
+                self.manager_rx.next().await
+            } else {
+                self.checked_start(batch).await;
+
+                // Include a timeout so we can start the next batch of
+                // shards even if no more messages are received.
+                timeout(self.wait_time_between_shard_start, self.manager_rx.next())
+                    .await
+                    .unwrap_or_default()
+            };
```

Comment:
What's the reason for having two separate branches? Yeah, adding a timeout only makes sense if there are shards in the queue waiting to get started (since we don't want to block forever), but I think just swapping the order serves well enough here.

Empty batches are handled fine by `checked_start`, so if multiple timeouts happen while waiting for a shard to get queued, all that will happen is a few extra empty `Vec` will get created.

Reply:
I just don't like calling `self.manager_rx.next()` in two different places. Adding a timer to the future really introduces very little overhead (what with zero-cost futures and all that, plus tokio relies on the system timer and doesn't just spinlock).

If you'd like to clean up the extra Vec usage, then you could either inline `checked_start` into the loop, or change `ShardQueue::pop_batch` to return `Option<Vec<ShardId>>` (`None` for empty batches), or both.

Reply:
My problem isn't the Vec or the timer future itself, it's CPU time spent on even running the loop when it's known that it won't do anything, especially if the user overrides the wait time. The loop could end up pretty much just spinning if the wait time is set to zero (minus tokio yielding due to coop budget).

Keeping `checked_start` as its own function seems cleaner to me, but I can change `pop_batch` to return `Option`. That seems like it improves the intent.

Reply:
Tried to verify if my concerns about this are actually measurable and the answer is not really. I'll just revert this snippet to what it was before and avoid any further changes.